My Project/clsTranslateWinForms.vb

```diff
@@ -1 +1,638 @@
+
+            'in the translation text, convert any single quotes to make them suitable for the SQL command
+            strText = strText.Replace("'", "''")
+
+            'get all translations for the specified form and language
+            'Note: The second `SELECT` is needed because we may sometimes need to translate  
+            '      translated text back to the original text (e.g. from French to English when 
+            '      the dialog language toggle button is clicked).
+            clsCommand.CommandText = "SELECT translation FROM translations WHERE language_code = '" &
+                                     strLanguage & "' AND id_text = '" & strText & "' OR (language_code = '" &
+                                     strLanguage & "' AND id_text = " &
+                                     "(SELECT id_text FROM translations WHERE translation = '" & strText & "'))"
+            Dim clsReader As SQLiteDataReader = clsCommand.ExecuteReader()
+            Using clsReader
+                'for each translation row
+                While (clsReader.Read())
+                    'ignore rows where the translation text is null or missing
+                    If clsReader.FieldCount < 1 OrElse clsReader.IsDBNull(0) Then
+                        Continue While
+                    End If
+                    'return the translation text
+                    Return clsReader.GetString(0)
+                End While
+            End Using
+        End Using
+        'if no tranlsation text was found then return original text unchanged
+        Return strText
+    End Function
+
+    '''--------------------------------------------------------------------------------------------
+    ''' <summary>   
+    '''    Decides whether <paramref name="strText"/> is likely to be changed during execution of 
+    '''    the software. If no, then returns <paramref name="strText"/>. If yes, then returns 
+    '''    'ReplaceWithDynamicTranslation'. It makes the decision based upon a set of heuristics.
+    '''    <para>
+    '''    This function is normally only used when creating a comma-separated list suitable for 
+    '''    importing into a database. During program execution, the 'ReplaceWithDynamicTranslation'
+    '''    text tells the library to dynamically try and translate the current text, rather than
+    '''    looking up the static text associated with the control.</para></summary>
+    '''
+    ''' <param name="strText">  The text to assess. </param>
+    '''
+    ''' <returns>   Decides whether <paramref name="strText"/> is likely to be changed during 
+    '''             execution of the software. If no, then returns <paramref name="strText"/>. 
+    '''             If yes, then returns'ReplaceWithDynamicTranslation'. </returns>
+    '''--------------------------------------------------------------------------------------------
+    Private Shared Function GetCsvText(strText As String) As String
```

Comment:
@lloyddewit if this function has to be in this library yet used across different products, these heuristics will have to be defined at the product level. For instance most of the heuristics statically typed here don't apply to climsoft.

Reply:
That's a very good point.
I suggest that we:

- Keep the check for multiline text,
- Keep the check for text that doesn't contain any letters
- Move the other checks to the product level

For 'doNotTranslate' we have the ignore file. Should we also have a similar txt file that specifies the patterns to translate dynamically? There are also other possible approaches.
What would work best for Climsoft?
We can discuss.

Reply:
I opened issue #4 for this.